analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/LocksReleasedAllPathsBase.cs

```diff
@@ -156,5 +162,52 @@ private ProgramState RemoveLock(SymbolicContext context, ISymbol symbol)
 
         private static ISymbol FirstArgumentSymbol(IInvocationOperationWrapper invocation) =>
             IArgumentOperationWrapper.FromOperation(invocation.Arguments.First()).Value.TrackedSymbol();
+
+        protected sealed class LockAcquireReleaseCollector
+        {
+            private static readonly HashSet<string> LockTypes = new() { "Mutex" }; // For some APIs ctor can directly acquire the lock (e.g. Mutex).
+
+            private static readonly HashSet<string> LockMethods = new(ReaderWriterLockSlimLockMethods)
+            {
+                "AcquireReaderLock",
+                "AcquireWriterLock",
+                "Enter",
+                "TryEnter",
+                "WaitOne"
+            };
+
+            private static readonly HashSet<string> ReleaseMethods = new()
+            {
+                "Exit",
+                "ExitReadLock",
+                "ExitUpgradeableReadLock",
+                "ExitWriteLock",
+                "ReleaseLock",
+                "ReleaseMutex",
+                "ReleaseReaderLock",
+                "ReleaseWriterLock"
+            };
+
+            private bool lockAcquired;
+            private bool lockReleased;
+
+            public bool LockAcquiredAndReleased =>
+                lockAcquired && lockReleased;
+
+            public void RegisterIdentifier(string name)
+            {
+                lockAcquired |= IsLockType(name) || IsLockMethod(name);
```

Comment:
This is bitwise operator, so it evaluates `IsLockType` and `IsLockMethod` even when we already know that `lockAcquired` is true. 
```suggestion
                lockAcquired = lockAcquired || IsLockType(name) || IsLockMethod(name);
```